File: adapters/anthropic.py
"""
Anthropic Messages API 协议兼容适配器。

接收 Anthropic Messages API 格式请求 (/v1/messages)，
转换为 OpenAI chat/completions 格式发送到 Copilot API，
再将 Copilot 响应转换回 Anthropic 格式。
"""

import json
from typing import Any
import uuid

import requests
from flask import Response, request

from adapters.base import BaseAdapter
from adapters.common import clean_body, get_fallback_model, is_model_not_supported, set_fallback_model, try_fallback
from auth import get_copilot_token
from config import COPILOT_API_BASE
from proxy import build_headers, forward_request

STREAM_END_MARKER = 'data: [DONE]\n\n'

FINISH_REASON_MAP: dict[str | None, str] = {
    'stop': 'end_turn',
    'length': 'max_tokens',
    'content_filter': 'content_filter',
}

ROLE_MAP: dict[str, str] = {
    'user': 'user',
    'assistant': 'assistant',
}


def _convert_anthropic_messages(anthropic_msg: list[dict[str, Any]]) -> list[dict[str, Any]]:
    """Anthropic messages → OpenAI messages (role/content pairs)."""
    result = []
    for msg in anthropic_msg:
        role = msg.get('role', 'user')
        content = msg.get('content', '')
        if isinstance(content, list):
            texts = []
            for block in content:
                if isinstance(block, dict) and block.get('type') == 'text':
                    texts.append(block.get('text', ''))
            content = '\n'.join(texts)
        result.append({'role': ROLE_MAP.get(role, role), 'content': content})
    return result


def anthropic_to_openai(payload: dict[str, Any]) -> dict[str, Any]:
    """将 Anthropic Messages API 请求转换为 OpenAI chat/completions 格式。"""
    messages = []

    system = payload.get('system')
    if system:
        messages.append({'role': 'system', 'content': system})

    anthropic_msgs = payload.get('messages', [])
    messages.extend(_convert_anthropic_messages(anthropic_msgs))

    oai_payload: dict[str, Any] = {
        'model': payload.get('model', ''),
        'messages': messages,
        'max_tokens': payload.get('max_tokens', 4096),
        'stream': payload.get('stream', False),
    }

    if 'temperature' in payload:
        oai_payload['temperature'] = payload['temperature']
    if 'top_p' in payload:
        oai_payload['top_p'] = payload['top_p']
    if 'stop_sequences' in payload:
        oai_payload['stop'] = payload['stop_sequences']
    if 'metadata' in payload:
        oai_payload['metadata'] = payload['metadata']

    return oai_payload


def _map_finish_reason(oai_reason: str | None) -> str:
    return FINISH_REASON_MAP.get(oai_reason, 'end_turn')


def openai_to_anthropic_response(oai_data: dict[str, Any]) -> dict[str, Any]:
    """将 OpenAI chat/completions 响应转换为 Anthropic Messages API 格式。"""
    choices = oai_data.get('choices', [{}])
    choice = choices[0] if choices else {}
    message = choice.get('message', {})
    content_text = message.get('content', '') or ''
    finish = _map_finish_reason(choice.get('finish_reason'))

    oai_usage = oai_data.get('usage', {}) or {}
    anthropic_usage = {
        'input_tokens': oai_usage.get('prompt_tokens', 0),
        'output_tokens': oai_usage.get('completion_tokens', 0),
    }

    return {
        'id': f"msg_{oai_data.get('id', str(uuid.uuid4().hex[:24]))}",
        'type': 'message',
        'role': 'assistant',
        'content': [{'type': 'text', 'text': content_text}],
        'model': oai_data.get('model', ''),
        'stop_reason': finish,
        'stop_sequence': None,
        'usage': anthropic_usage,
    }


def _build_anthropic_usage(prompt_tokens: int = 0, completion_tokens: int = 0) -> dict[str, Any]:
    return {'input_tokens': prompt_tokens, 'output_tokens': completion_tokens}


def openai_sse_to_anthropic_sse() -> Response:
    """将 OpenAI SSE 流实时转换为 Anthropic SSE 事件流。"""
    url = request.url.replace('/v1/messages', '/v1/chat/completions')
    copilot_url = f"{COPILOT_API_BASE}/chat/completions"

    headers = build_headers(request.headers.get('Content-Type', 'application/json'))
    raw_body = clean_body(request.get_data())
    body_str = raw_body.decode('utf-8') if raw_body else '{}'
    oai_body = anthropic_to_openai(json.loads(body_str))
    oai_body['stream'] = True
    new_body = json.dumps(oai_body).encode()

    upstream = forward_request('POST', copilot_url, headers, new_body, stream=True)

    if upstream.status_code == 400 and is_model_not_supported(upstream):
        fb = try_fallback('POST', copilot_url, headers, new_body)
        if fb is not None:
            upstream = fb

    if upstream.status_code != 200:
        error_text = upstream.text[:500]
        print(f"[Anthropic] API 返回 {upstream.status_code}: {error_text}")
        return _anthropic_error_response(upstream.status_code, error_text)

    def generate():
        message_id = f"msg_{uuid.uuid4().hex[:24]}"
        model_name = oai_body.get('model', '')
        has_started = False
        prompt_tokens = 0
        completion_tokens = 0
        stop_reason = None

        for chunk_bytes in upstream.iter_content(chunk_size=1024):
            if not chunk_bytes:
                continue

            for line in chunk_bytes.decode('utf-8', errors='replace').splitlines():
                line = line.strip()
                if not line or line == STREAM_END_MARKER.strip():
                    continue
                if not line.startswith('data: '):
                    continue

                data_str = line[6:]
                if data_str == '[DONE]':
                    break

                try:
                    chunk = json.loads(data_str)
                except json.JSONDecodeError:
                    continue

                choices = chunk.get('choices', [{}])
                delta = choices[0].get('delta', {}) if choices else {}
                finish = choices[0].get('finish_reason') if choices else None

                if not has_started:
                    has_started = True
                    usage = chunk.get('usage', {}) or {}
                    prompt_tokens = usage.get('prompt_tokens', 0)
                    completion_tokens = usage.get('completion_tokens', 0)

                    msg_start = {
                        'type': 'message_start',
                        'message': {
                            'id': message_id,
                            'type': 'message',
                            'role': 'assistant',
                            'content': [],
                            'model': model_name,
                            'stop_reason': None,
                            'stop_sequence': None,
                            'usage': _build_anthropic_usage(prompt_tokens, 0),
                        },
                    }
                    yield f"event: message_start\ndata: {json.dumps(msg_start)}\n\n"

                    content_block = {'type': 'text', 'text': ''}
                    cb_start = {
                        'type': 'content_block_start',
                        'index': 0,
                        'content_block': content_block,
                    }
                    yield f"event: content_block_start\ndata: {json.dumps(cb_start)}\n\n"

                text = delta.get('content', '')
                if text:
                    cb_delta = {
                        'type': 'content_block_delta',
                        'index': 0,
                        'delta': {'type': 'text_delta', 'text': text},
                    }
                    yield f"event: content_block_delta\ndata: {json.dumps(cb_delta)}\n\n"

                if finish:
                    stop_reason = _map_finish_reason(finish)
                    yield f"event: content_block_stop\ndata: {json.dumps({'type': 'content_block_stop', 'index': 0})}\n\n"

                    msg_delta = {
                        'type': 'message_delta',
                        'delta': {'stop_reason': stop_reason, 'stop_sequence': None},
                        'usage': _build_anthropic_usage(0, completion_tokens),
                    }
                    yield f"event: message_delta\ndata: {json.dumps(msg_delta)}\n\n"

                    yield f"event: message_stop\ndata: {json.dumps({'type': 'message_stop'})}\n\n"

        if not has_started:
            yield _anthropic_sse_error(503, '上游无响应')

    return Response(
        generate(),
        mimetype='text/event-stream',
        headers={
            'Cache-Control': 'no-cache',
            'X-Accel-Buffering': 'no',
        },
    )


def _anthropic_error_response(status_code: int, message: str) -> Response:
    body = {
        'type': 'error',
        'error': {
            'type': 'api_error',
            'message': message,
        },
    }
    return Response(json.dumps(body), status=status_code, mimetype='application/json')


def _anthropic_sse_error(status_code: int, message: str) -> str:
    err = {
        'type': 'error',
        'error': {'type': 'api_error', 'message': message},
    }
    return f"event: error\ndata: {json.dumps(err)}\n\n"


class AnthropicAdapter(BaseAdapter):

    def handle_request(self, path: str) -> Response | tuple[dict[str, Any], int]:
        if get_copilot_token() is None:
            return {'error': 'Copilot token 未就绪，请检查授权状态'}, 503

        try:
            payload = request.get_json(force=True)
        except Exception:
            return {'error': '无效的 JSON 请求体'}, 400

        if not payload or 'messages' not in payload:
            return {'error': '缺少必填字段: messages'}, 400

        stream = payload.get('stream', False)

        if stream:
            return openai_sse_to_anthropic_sse()

        oai_payload = anthropic_to_openai(payload)
        copilot_url = f'{COPILOT_API_BASE}/chat/completions'
        headers = build_headers('application/json')
        body = json.dumps(oai_payload).encode()

        try:
            resp = forward_request('POST', copilot_url, headers, body)

            if resp.status_code == 400 and is_model_not_supported(resp):
                fb = try_fallback('POST', copilot_url, headers, body)
                if fb is not None:
                    resp = fb

            if resp.status_code != 200:
                error_text = resp.text[:500]
                print(f'[Anthropic] API 返回 {resp.status_code}: {error_text}')
                sc: int = resp.status_code
                return _anthropic_error_response(sc, error_text)

            oai_data = resp.json()
            anthropic_resp = openai_to_anthropic_response(oai_data)
            return Response(json.dumps(anthropic_resp), mimetype='application/json')
        except requests.exceptions.Timeout:
            return {'error': '请求超时'}, 504
        except Exception as e:
            print(f'[✗] Anthropic 代理错误: {e}')
            return {'error': str(e)}, 502

File: internal/proxy/anthropic.go
package proxy

import (
	"bufio"
	"encoding/json"
	"fmt"
	"net/http"
	"strings"
	"time"
)

var anthropicFinishReasonMap = map[string]string{
	"stop":           "end_turn",
	"length":         "max_tokens",
	"content_filter": "content_filter",
}

var anthropicModelMap = map[string]string{
	"claude-opus-4-7":            "claude-opus-4.7",
	"claude-opus-4.7":            "claude-opus-4.7",
	"claude-sonnet-4-6":          "claude-sonnet-4.6",
	"claude-sonnet-4.6":          "claude-sonnet-4.6",
	"claude-opus-4-6":            "claude-opus-4.6",
	"claude-opus-4.6":            "claude-opus-4.6",
	"claude-haiku-4-5":           "claude-haiku-4.5",
	"claude-haiku-4.5":           "claude-haiku-4.5",
	"claude-haiku-4-5-20251001":  "claude-haiku-4.5",
	"claude-sonnet-4-5":          "claude-sonnet-4.5",
	"claude-sonnet-4.5":          "claude-sonnet-4.5",
	"claude-sonnet-4-5-20250929": "claude-sonnet-4.5",
	"claude-opus-4-5":            "claude-opus-4.5",
	"claude-opus-4.5":            "claude-opus-4.5",
	"claude-opus-4-5-20251101":   "claude-opus-4.5",
	"claude-opus-4-1":            "claude-opus-4.1",
	"claude-opus-4.1":            "claude-opus-4.1",
	"claude-opus-4-1-20250805":   "claude-opus-4.1",
	"claude-sonnet-4-0":          "claude-sonnet-4.0",
	"claude-sonnet-4.0":          "claude-sonnet-4.0",
	"claude-sonnet-4-20250514":   "claude-sonnet-4.0",
	"claude-opus-4-0":            "claude-opus-4.0",
	"claude-opus-4.0":            "claude-opus-4.0",
	"claude-opus-4-20250514":     "claude-opus-4.0",
	"claude-3-7-sonnet-latest":   "claude-3.7-sonnet",
	"claude-3-7-sonnet-20250219": "claude-3.7-sonnet",
	"claude-3.7-sonnet":          "claude-3.7-sonnet",
	"claude-3-5-sonnet-latest":   "claude-3.5-sonnet",
	"claude-3-5-sonnet-20241022": "claude-3.5-sonnet",
	"claude-3-5-sonnet-20240620": "claude-3.5-sonnet",
	"claude-3.5-sonnet":          "claude-3.5-sonnet",
	"claude-3-5-haiku-latest":    "claude-3.5-haiku",
	"claude-3-5-haiku-20241022":  "claude-3.5-haiku",
	"claude-3-5-haiku":           "claude-3.5-haiku",
	"claude-3.5-haiku":           "claude-3.5-haiku",
	"claude-3-opus-latest":       "claude-3-opus",
	"claude-3-opus-20240229":     "claude-3-opus",
	"claude-3-opus":              "claude-3-opus",
	"claude-3-sonnet-20240229":   "claude-3-sonnet",
	"claude-3-sonnet":            "claude-3-sonnet",
	"claude-3-haiku-20240307":    "claude-3-haiku",
	"claude-3-haiku":             "claude-3-haiku",
}

func (h *Handler) ServeAnthropicMessages(w http.ResponseWriter, r *http.Request) {
	start := time.Now()
	if r.Method != http.MethodPost {
		w.Header().Set("Allow", http.MethodPost)
		writeJSON(w, http.StatusMethodNotAllowed, map[string]string{"error": "method not allowed"})
		h.record(RequestRecord{Time: start, Protocol: "anthropic", Method: r.Method, Path: r.URL.Path, Status: http.StatusMethodNotAllowed, DurationMs: time.Since(start).Milliseconds(), Error: "method not allowed"})
		return
	}
	if h.proxyDisabled() {
		writeJSON(w, http.StatusServiceUnavailable, map[string]string{"error": "proxy service is disabled"})
		h.record(RequestRecord{Time: start, Protocol: "anthropic", Method: r.Method, Path: r.URL.Path, Status: http.StatusServiceUnavailable, DurationMs: time.Since(start).Milliseconds(), Error: "proxy disabled"})
		return
	}
	if !h.authorized(r) {
		writeJSON(w, http.StatusUnauthorized, map[string]string{"error": "invalid api key"})
		h.record(RequestRecord{Time: start, Protocol: "anthropic", Method: r.Method, Path: r.URL.Path, Status: http.StatusUnauthorized, DurationMs: time.Since(start).Milliseconds(), Error: "invalid api key"})
		return
	}

	token := h.auth.CopilotToken()
	if token == "" {
		writeJSON(w, http.StatusServiceUnavailable, map[string]string{"error": "Copilot token 未就绪，请检查授权状态"})
		h.record(RequestRecord{Time: start, Protocol: "anthropic", Method: r.Method, Path: r.URL.Path, Status: http.StatusServiceUnavailable, DurationMs: time.Since(start).Milliseconds(), Error: "missing copilot token"})
		return
	}

	var payload map[string]any
	r.Body = http.MaxBytesReader(w, r.Body, 10<<20)
	if err := json.NewDecoder(r.Body).Decode(&payload); err != nil {
		writeJSON(w, http.StatusBadRequest, map[string]string{"error": "无效的 JSON 请求体"})
		h.record(RequestRecord{Time: start, Protocol: "anthropic", Method: r.Method, Path: r.URL.Path, Status: http.StatusBadRequest, DurationMs: time.Since(start).Milliseconds(), Error: "invalid json"})
		return
	}
	if _, ok := payload["messages"]; !ok {
		writeJSON(w, http.StatusBadRequest, map[string]string{"error": "缺少必填字段: messages"})
		h.record(RequestRecord{Time: start, Protocol: "anthropic", Method: r.Method, Path: r.URL.Path, Model: stringValue(payload["model"]), Status: http.StatusBadRequest, DurationMs: time.Since(start).Milliseconds(), Error: "missing messages"})
		return
	}

	model := stringValue(payload["model"])
	if !h.checkModelAllowed(model) {
		writeJSON(w, http.StatusForbidden, map[string]string{"error": ModelNotAllowedError(model)})
		h.record(RequestRecord{Time: start, Protocol: "anthropic", Method: r.Method, Path: r.URL.Path, Model: model, Status: http.StatusForbidden, DurationMs: time.Since(start).Milliseconds(), Error: ModelNotAllowedError(model)})
		return
	}

	oaiPayload, err := anthropicToOpenAI(payload)
	if err != nil {
		writeJSON(w, http.StatusBadRequest, map[string]string{"error": err.Error()})
		h.record(RequestRecord{Time: start, Protocol: "anthropic", Method: r.Method, Path: r.URL.Path, Model: model, Status: http.StatusBadRequest, DurationMs: time.Since(start).Milliseconds(), Error: err.Error()})
		return
	}
	if stream, _ := payload["stream"].(bool); stream {
		oaiPayload["stream"] = true
		ensureStreamUsage(oaiPayload)
		h.serveAnthropicStream(w, r, token, oaiPayload, start)
		return
	}

	body, err := json.Marshal(oaiPayload)
	if err != nil {
		writeJSON(w, http.StatusBadRequest, map[string]string{"error": err.Error()})
		h.record(RequestRecord{Time: start, Protocol: "anthropic", Method: r.Method, Path: r.URL.Path, Model: stringValue(oaiPayload["model"]), Status: http.StatusBadRequest, DurationMs: time.Since(start).Milliseconds(), Error: err.Error()})
		return
	}
	copilotURL := h.chatCompletionsURL()
	resp, err := h.forward(r.Context(), http.MethodPost, copilotURL, token, "application/json", body)
	if err != nil {
		writeJSON(w, http.StatusBadGateway, map[string]string{"error": err.Error()})
		h.record(RequestRecord{Time: start, Protocol: "anthropic", Method: r.Method, Path: r.URL.Path, Model: stringValue(oaiPayload["model"]), Status: http.StatusBadGateway, DurationMs: time.Since(start).Milliseconds(), Error: err.Error()})
		return
	}
	defer resp.Body.Close()

	if resp.StatusCode != http.StatusOK {
		body := readErrorPreview(resp)
		anthropicError(w, resp.StatusCode, body)
		h.record(RequestRecord{Time: start, Protocol: "anthropic", Method: r.Method, Path: r.URL.Path, Model: stringValue(oaiPayload["model"]), Status: resp.StatusCode, DurationMs: time.Since(start).Milliseconds(), Error: body})
		return
	}

	var oaiData map[string]any
	if err := json.NewDecoder(resp.Body).Decode(&oaiData); err != nil {
		writeJSON(w, http.StatusBadGateway, map[string]string{"error": err.Error()})
		h.record(RequestRecord{Time: start, Protocol: "anthropic", Method: r.Method, Path: r.URL.Path, Model: stringValue(oaiPayload["model"]), Status: http.StatusBadGateway, DurationMs: time.Since(start).Milliseconds(), Error: err.Error()})
		return
	}
	record := RequestRecord{Time: start, Protocol: "anthropic", Method: r.Method, Path: r.URL.Path, Model: stringValue(oaiPayload["model"]), Status: http.StatusOK, DurationMs: time.Since(start).Milliseconds()}
	if model := stringValue(oaiData["model"]); model != "" {
		record.Model = model
	}
	if usage, ok := oaiData["usage"].(map[string]any); ok {
		record.PromptTokens, record.CompletionTokens, record.TotalTokens = tokensFromUsage(usage)
	}
	h.record(record)
	writeJSON(w, http.StatusOK, openAIToAnthropicResponse(oaiData))
}

func anthropicToOpenAI(payload map[string]any) (map[string]any, error) {
	model, err := normalizeAnthropicModel(stringValue(payload["model"]))
	if err != nil {
		return nil, err
	}

	messages := make([]map[string]any, 0)
	if system, ok := payload["system"]; ok && system != nil {
		text, err := anthropicContentToText(system)
		if err != nil {
			return nil, err
		}
		if text != "" {
			messages = append(messages, map[string]any{"role": "system", "content": text})
		}
	}
	rawMessages, ok := payload["messages"].([]any)
	if !ok {
		return nil, fmt.Errorf("invalid Anthropic messages: expected array, got %T", payload["messages"])
	}
	for _, item := range rawMessages {
		msg, ok := item.(map[string]any)
		if !ok {
			return nil, fmt.Errorf("invalid Anthropic message: expected object, got %T", item)
		}
		role, _ := msg["role"].(string)
		if role == "" {
			role = "user"
		}
		// 角色映射：Anthropic "assistant" → OpenAI "assistant"，其余映射为 "user"
		// 之前 Bug: 所有非 assistant 角色（包括正确的 "user"）都被强制映射成了 "user"，没有走这个分支
		if role != "user" && role != "assistant" {
			role = "user"
		}
		content, err := anthropicContentToText(msg["content"])
		if err != nil {
			return nil, err
		}
		messages = append(messages, map[string]any{
			"role":    role,
			"content": content,
		})
	}

	result := map[string]any{
		"model":      model,
		"messages":   messages,
		"max_tokens": valueOrDefault(payload, "max_tokens", float64(4096)),
		"stream":     valueOrDefault(payload, "stream", false),
	}
	copyIfPresent(result, payload, "temperature", "temperature")
	copyIfPresent(result, payload, "top_p", "top_p")
	copyIfPresent(result, payload, "stop_sequences", "stop")
	copyIfPresent(result, payload, "metadata", "metadata")
	return result, nil
}

func normalizeAnthropicModel(model string) (string, error) {
	if extractVendor(model) != "" {
		return model, nil
	}
	canonical, ok := anthropicModelMap[strings.ToLower(model)]
	if !ok {
		return "", fmt.Errorf("unsupported Anthropic model: %s", model)
	}
	return canonical, nil
}

func anthropicContentToText(content any) (string, error) {
	switch v := content.(type) {
	case string:
		return v, nil
	case []any:
		texts := make([]string, 0, len(v))
		for _, item := range v {
			block, ok := item.(map[string]any)
			if !ok {
				return "", fmt.Errorf("invalid Anthropic content block: expected object, got %T", item)
			}
			blockType, _ := block["type"].(string)
			if blockType != "text" {
				return "", fmt.Errorf("unsupported Anthropic content type: %s", blockType)
			}
			textVal, ok := block["text"].(string)
			if !ok {
				return "", fmt.Errorf("invalid Anthropic text block: 'text' field is missing or not a string")
			}
			texts = append(texts, textVal)
		}
		return strings.Join(texts, "\n"), nil
	default:
		return "", fmt.Errorf("invalid Anthropic content: expected string or array, got %T", v)
	}
}

func anthropicUsageFromOpenAIUsage(usage map[string]any) map[string]any {
	return map[string]any{
		"input_tokens":  orZero(usage["prompt_tokens"]),
		"output_tokens": orZero(usage["completion_tokens"]),
	}
}

func openAIToAnthropicResponse(oaiData map[string]any) map[string]any {
	choice := firstChoice(oaiData)
	message, _ := choice["message"].(map[string]any)
	content := stringValue(message["content"])
	finish := anthropicFinishReason(stringValue(choice["finish_reason"]))
	usage, _ := oaiData["usage"].(map[string]any)

	id := stringValue(oaiData["id"])
	if id == "" {
		id = randomHex(24)
	}

	return map[string]any{
		"id":            "msg_" + strings.TrimPrefix(id, "msg_"),
		"type":          "message",
		"role":          "assistant",
		"content":       []map[string]any{{"type": "text", "text": content}},
		"model":         stringValue(oaiData["model"]),
		"stop_reason":   finish,
		"stop_sequence": nil,
		"usage":         anthropicUsageFromOpenAIUsage(usage),
	}
}

func (h *Handler) serveAnthropicStream(w http.ResponseWriter, r *http.Request, token string, oaiPayload map[string]any, start time.Time) {
	body, err := json.Marshal(oaiPayload)
	if err != nil {
		writeJSON(w, http.StatusBadRequest, map[string]string{"error": err.Error()})
		h.record(RequestRecord{Time: start, Protocol: "anthropic", Method: r.Method, Path: r.URL.Path, Model: stringValue(oaiPayload["model"]), Status: http.StatusBadRequest, DurationMs: time.Since(start).Milliseconds(), Error: err.Error()})
		return
	}
	copilotURL := h.chatCompletionsURL()
	resp, err := h.forward(r.Context(), http.MethodPost, copilotURL, token, "application/json", body)
	if err != nil {
		writeJSON(w, http.StatusBadGateway, map[string]string{"error": err.Error()})
		h.record(RequestRecord{Time: start, Protocol: "anthropic", Method: r.Method, Path: r.URL.Path, Model: stringValue(oaiPayload["model"]), Status: http.StatusBadGateway, DurationMs: time.Since(start).Milliseconds(), Error: err.Error()})
		return
	}
	defer resp.Body.Close()

	if resp.StatusCode != http.StatusOK {
		body := readErrorPreview(resp)
		anthropicError(w, resp.StatusCode, body)
		h.record(RequestRecord{Time: start, Protocol: "anthropic", Method: r.Method, Path: r.URL.Path, Model: stringValue(oaiPayload["model"]), Status: resp.StatusCode, DurationMs: time.Since(start).Milliseconds(), Error: body})
		return
	}

	w.Header().Set("Content-Type", "text/event-stream; charset=utf-8")
	w.Header().Set("Cache-Control", "no-cache")
	w.Header().Set("X-Accel-Buffering", "no")
	flusher, _ := w.(http.Flusher)

	messageID := "msg_" + randomHex(24)
	model := stringValue(oaiPayload["model"])
	started := false
	streamRecord := RequestRecord{Time: start, Protocol: "anthropic", Method: r.Method, Path: r.URL.Path, Model: model}
	scanner := bufio.NewScanner(resp.Body)
	scanner.Buffer(make([]byte, 1024), 1024*1024)
	for scanner.Scan() {
		line := strings.TrimSpace(scanner.Text())
		if line == "" || !strings.HasPrefix(line, "data: ") {
			continue
		}
		data := strings.TrimPrefix(line, "data: ")
		if data == "[DONE]" {
			break
		}
		var chunk map[string]any
		if err := json.Unmarshal([]byte(data), &chunk); err != nil {
			continue
		}
		mergeStreamUsage(&streamRecord, chunk)
		if model := stringValue(chunk["model"]); model != "" {
			streamRecord.Model = model
		}
		if !started {
			started = true
			writeSSE(w, "message_start", map[string]any{
				"type": "message_start",
				"message": map[string]any{
					"id":            messageID,
					"type":          "message",
					"role":          "assistant",
					"content":       []any{},
					"model":         model,
					"stop_reason":   nil,
					"stop_sequence": nil,
					"usage":         map[string]any{"input_tokens": 0, "output_tokens": 0},
				},
			})
			writeSSE(w, "content_block_start", map[string]any{
				"type":          "content_block_start",
				"index":         0,
				"content_block": map[string]any{"type": "text", "text": ""},
			})
			flush(flusher)
		}
		choice := firstChoice(chunk)
		delta, _ := choice["delta"].(map[string]any)
		if text := stringValue(delta["content"]); text != "" {
			writeSSE(w, "content_block_delta", map[string]any{
				"type":  "content_block_delta",
				"index": 0,
				"delta": map[string]any{"type": "text_delta", "text": text},
			})
			flush(flusher)
		}
		if finish := stringValue(choice["finish_reason"]); finish != "" {
			writeSSE(w, "content_block_stop", map[string]any{"type": "content_block_stop", "index": 0})
			// 从上游 OpenAI chunk 中提取 usage 而非硬编码为 0
			usage := map[string]any{"prompt_tokens": streamRecord.PromptTokens, "completion_tokens": streamRecord.CompletionTokens}
			if streamRecord.PromptTokens == 0 && streamRecord.CompletionTokens == 0 {
				// 如果流中没有 usage 数据，检查最后一个 chunk 是否有 usage
				if chunkUsage, ok := chunk["usage"].(map[string]any); ok {
					if pt, ct, _ := tokensFromUsage(chunkUsage); pt > 0 || ct > 0 {
						usage = chunkUsage
					}
				}
			}
			writeSSE(w, "message_delta", map[string]any{
				"type":  "message_delta",
				"delta": map[string]any{"stop_reason": anthropicFinishReason(finish), "stop_sequence": nil},
				"usage": anthropicUsageFromOpenAIUsage(usage),
			})
			writeSSE(w, "message_stop", map[string]any{"type": "message_stop"})
			flush(flusher)
		}
	}
	if !started {
		writeSSE(w, "error", map[string]any{"type": "error", "error": map[string]any{"type": "api_error", "message": "上游无响应"}})
	}
	status := http.StatusOK
	errMsg := ""
	if !started {
		status = http.StatusServiceUnavailable
		errMsg = "empty upstream stream"
	}
	streamRecord.Status = status
	streamRecord.DurationMs = time.Since(start).Milliseconds()
	streamRecord.Error = errMsg
	h.record(streamRecord)
}

func anthropicFinishReason(reason string) string {
	if mapped, ok := anthropicFinishReasonMap[reason]; ok {
		return mapped
	}
	return "end_turn"
}

func anthropicError(w http.ResponseWriter, status int, message string) {
	writeJSON(w, status, map[string]any{
		"type": "error",
		"error": map[string]any{
			"type":    "api_error",
			"message": message,
		},
	})
}

func writeSSE(w http.ResponseWriter, event string, payload any) {
	raw, _ := json.Marshal(payload)
	fmt.Fprintf(w, "event: %s\ndata: %s\n\n", event, raw)
}
